### DISCUSSION TITLE
Crosscompilation of FMUs on Windows not working (triggered from OMEdit)

### DISCUSSION THREAD

**Discussion by dietmarw** [OpenModelica/OpenModelica#10411] [Q&A]
Hi,
I have a simple demo model of a sinsource
[SinSource.mo.zip](https://github.com/OpenModelica/OpenModelica/files/10991207/SinSource.mo.zip)
 that I try to export as a cs-FMU for arm32 target. 
Doing so using OMEdit on Linux works without problems. Now trying the same with OMEdit v1.20.0 on Windows gives me: 

```
[5] 13:58:47 Translation Error
Error building simulator. Build log: docker run  --rm -w /fmu -v 3d44937f1948114d227158413e2d85ced9d29e96d72e85eba28675892c8305f2:/fmu --pull=never multiarch/crossbuild sh -c "cd "/fmu/sine.fmutmp/sources" && ./configure --host='arm-linux-gnueabihf' CFLAGS='-Os' CPPFLAGS=-I/fmu/fmiInclude LDFLAGS= && mingw32-make -j6 nozip":
3d44937f1948114d227158413e2d85ced9d29e96d72e85eba28675892c8305f2
sh: 1: ./configure: not found
```

I made sure that docker is installed and that the mutiarch/crossbuild image has been pulled beforehand. So something seems to be missing. Any suggestions?

**Comment by adeas31** [comment]
@AnHeuermann can you please check that?

**Reply by AnHeuermann** [reply]
The Docker cross-compilation from Windows should automatically use the CMake FMU build. I'm not sure why this isn't the case here.

Can you try with the CMake option enabled?
But I can look into this next week.

**Comment by dietmarw** [comment]
I can try. How/where can I enable the CMake option?

**Reply by AnHeuermann** [reply]
It is [--fmuCMakeBuild=true](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-fmucmakebuild). I guess you disabled it explicitly in your OMEdit settings (Options->Simulation->Additional Translation Flags).

When I run

```Modelica
loadModel(Modelica); getErrorString();
buildModelFMU(Modelica.Blocks.Sources.Sine, platforms={"arm-linux-gnueabihf docker run --pull=never multiarch/crossbuild"}); getErrorString();
```
or do the equivalent in OMEdit

Docker with CMake is used.

```bash
Translation Notification
OS is Windows and docker platform detected. Using CMake to build FMU.

Scripting Notification
The FMU is generated at /path/to/OpenModelica/OMEdit/Modelica.Blocks.Sources.Sine/Modelica.Blocks.Sources.Sine.fmu.
```

You also need to add [--fmuRuntimeDepends=none](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-fmuruntimedepends) to skip the CMake feature that needs a newer version. The image multiarch/crossbuild doesn't have a recent enough CMake version.

**Reply by dietmarw** [reply]
> It is [--fmuCMakeBuild=true](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-fmucmakebuild). I guess you disabled it explicitly in your OMEdit settings (Options->Simulation->Additional Translation Flags).

By default that field is empty and hence FMU generation using the cross-compilation options will fail (on Windows that is). Maybe this should be changed so "normal" Windows users don't have specify some (for them most likely) cryptic translation flags.

> You also need to add [--fmuRuntimeDepends=none](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-fmuruntimedepends) to skip the CMake feature that needs a newer version. The image multiarch/crossbuild doesn't have a recent enough CMake version.

With those two flags I got one step further but now I get loads of compilation errors: 

[FMU.log](https://github.com/OpenModelica/OpenModelica/files/11080625/FMU.log)

So something is still not right with the way OMEdit is setup on Windows.

**Reply by AnHeuermann** [reply]
> By default that field is empty and hence FMU generation using the cross-compilation options will fail (on Windows that is). Maybe this should be changed so "normal" Windows users don't have specify some (for them most likely) cryptic translation flags.

It should already be this way. If the flag value is not provided, but the OS is Windows and cross-compilation using Docker is detected (we search for `docker` in the `platforms` string) omc will use the CMake FMU build and for me it is working this way. Maybe the part where we detect if Docker is used is not working in your case.

> With those two flags I got one step further but now I get loads of compilation errors:

Can you add a minimal working example to reproduce this issue?
Does the script I added above work for you?

You have some non-default options enabled in OMEdit. In this case you are exporting a CS FMU using CVODE as ODE solver (default is Explicit Euler). There was a bug in omc where the linking was broken in this specific case. I fixed that one today: https://github.com/OpenModelica/OpenModelica/commit/02fef0b9470bc64dc4f0fc2694513ac72355c22a

**Reply by AnHeuermann** [reply]
Oh and I'm not sure if I tested 02fef0b on Windows. Maybe I can check it tomorrow.

**Reply by dietmarw** [reply]
> It should already be this way. If the flag value is not provided, but the OS is Windows and cross-compilation using Docker is detected (we search for `docker` in the `platforms` string) omc will use the CMake FMU build and for me it is working this way. Maybe the part where we detect if Docker is used is not working in your case.

Yes and it would be good if this could be fixed so it works out of the box. 
 
> > With those two flags I got one step further but now I get loads of compilation errors:
> 
> Can you add a minimal working example to reproduce this issue? Does the script I added above work for you?

I have. I've provided the mo file in the zip file. Running the command you provided throws a parsing error:

```
[1] 08:27:52 Scripting Error
Error parsing expression: readOperator got unknown operator {.
```

> You have some non-default options enabled in OMEdit. In this case you are exporting a CS FMU using CVODE as ODE solver (default is Explicit Euler). There was a bug in omc where the linking was broken in this specific case. I fixed that one today: [02fef0b](https://github.com/OpenModelica/OpenModelica/commit/02fef0b9470bc64dc4f0fc2694513ac72355c22a)

OK, I was not aware that Euler is the default (since the settings dialogue is not marking which setting is default) and the fields get easily changed whilst scrolling when the mouse is touching one of the fields. Anyway redone the test with Euler and still get errors:

[FMU2.txt](https://github.com/OpenModelica/OpenModelica/files/11086118/FMU2.txt)

**Reply by AnHeuermann** [reply]
Sorry, I can't reproduce your issue.
Your example is missing model `SinSource_me_FMU` and how to generate it:
```
[1] 14:07:58 Translation Error
[SinSource: 6:3-7:114]: [Class SinSource_me_FMU not found in scope SinSource.](omeditmessagesbrowser:///SinSource?lineNumber=6)
```
I assumed it is some model exchange 2.0 FMU that was included with [importFMU](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html#importfmu), but I don't know how you generated the FMU and from what Modelica model.

Please provide a full example and open a ticket about this. We can ignore OMEdit for now, it's an issue in the compiler itself. So I would suggest to use a mos script to make the example reproducible independent from your OMEdit settings.
You can run mos scripts with omc by starting OMEdit, than open a command prompt from `Tools->OpenModelica Command Prompt`. Navigate to your mos script and run it with:

```cmd
omc test.mos
```

And please add small Modelica files directly into the markdown of the issue, so we don't have to download random zip files:

```markdown
    ```modelica
    model M
    end M;
    ```
```

**Reply by dietmarw** [reply]
OK I think I stick with your minimal example, could you tell me point me to how/where to add the translation flag `--fmuRuntimeDepends=none` in the call below?

```
loadModel(Modelica); getErrorString();
buildModelFMU(Modelica.Blocks.Sources.Sine, platforms={"arm-linux-gnueabihf docker run --pull=never multiarch/crossbuild"}); getErrorString();
```

**Reply by dietmarw** [reply]
BTW, there seems still to be an issue with OMEdit here, since calling the above with omc will report 
"Notification: OS is Windows and docker platform detected. Using CMake to build FMU."
which does not work without the translation flag `--fmuCMakeBuild=true` in OMEdit. So somehow the docker detections is not triggered when starting the export from OMEdit

**Reply by AnHeuermann** [reply]
Sure. It is with [setCommandLineOptions](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/scripting_api.html#setcommandlineoptions). So in this case add

```modelica
setCommandLineOptions("--fmuRuntimeDepends=none");
getErrorString();  // Always add this after each call. Otherwise errors are not shown. Stupid, but that's the way it is.
```

at some point before calling `buildModelFMU`.

> So somehow the docker detections is not triggered when starting the export from OMEdit

Maybe @adeas31 can help with that one. I'm still assuming that in some location the setting `--fmuCMakeBuild=false` is specified in OMEdit. You could check your omeditcommunications.log (OMEdit Tools->Open Working Directory) and search for `fmuCMakeBuild`.
The location where one could set this in OMEdit would be in Tools->Options->Simulation in `Additional Translation Flags`. Or maybe it is because we are using different versions of omc, not sure.

![grafik](https://user-images.githubusercontent.com/38031952/228496916-40821278-0e9f-475f-ac41-0b45e3f08ce0.png)

**Reply by adeas31** [reply]
`fmuCMakeBuild` is not set by OMEdit which means it should use whatever is the compiler default for it.

**Comment by dietmarw** [comment]
So do you mean that those translation flags should be there by default? I'm mainly concerned about the "out of the box" experience for users (mostly my students). So if there is an option to export an FMU then it should just work and no messing about with translation flags in the settings. If this is needed then such translation flags should be added automatically when one selects a certain export option. Same applies to **not** having to pull a docker image manually ;-)

**Reply by AnHeuermann** [reply]
> So do you mean that those translation flags should be there by default?

No, those are just the one I happen to have in there (but `-d=backenddaeinfo` is very useful).

>  So if there is an option to export an FMU then it should just work and no messing about with translation flags in the settings. If this is needed then such translation flags should be added automatically when one selects a certain export option.

I agree. And in the nightly build this is somewhat working. Eventually we will make the CMake FMU export the default option. In the moment it should be superior to the Makefile build of FMUs, but still has issues. E.g. the dependency to CMake > v.3.21.
While your proposal from https://github.com/OpenModelica/OpenModelica/discussions/10459 could be a solution, it is also some work to maintain such a Docker image.

> Same applies to not having to pull a docker image manually ;-)

Maybe, but I already explained the reasoning behind that in https://github.com/OpenModelica/OpenModelica/issues/10174.

**Reply by dietmarw** [reply]
> I agree. And in the nightly build this is somewhat working. Eventually we will make the CMake FMU export the default option. In the moment it should be superior to the Makefile build of FMUs, but still has issues. E.g. the dependency to CMake > v.3.21. While your proposal from #10459 could be a solution, it is also some work to maintain such a Docker image.

I'd argue the maintenance effort would be rather minimal especially since this would also provide a solution to #10174 and give a better user experience of something that "just works".

> > Same applies to not having to pull a docker image manually ;-)
> 
> Maybe, but I already explained the reasoning behind that in #10174.

See above :-)
